lib/srv/db/server.go

```diff
@@ -807,6 +811,37 @@ func (s *Server) stopHeartbeat(name string) error {
 	return heartbeat.Close()
 }
 
+// stopUpstreamInventoryHeartbeat stops the upstream inventory control stream
+// heartbeat for a single database.
+// https://github.com/gravitational/teleport/issues/50237
+func (s *Server) stopUpstreamInventoryHeartbeat(ctx context.Context, db types.Database) error {
+	if _, ok := s.cfg.InventoryHandle.GetSender(); ok {
+		select {
+		// get latest sender
+		case sender := <-s.cfg.InventoryHandle.Sender():
+			if sender.Hello().Capabilities.DatabaseHeartbeatGracefulStop {
+				err := sender.Send(ctx, &proto.UpstreamInventoryStopHeartbeat{
+					Kind:   proto.StopHeartbeatKind_STOP_HEARTBEAT_KIND_DATABASE_SERVER,
+					HostID: s.cfg.HostID,
+					Name:   db.GetName(),
+				})
+				return trace.Wrap(err)
+			}
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		}
+	}
+	// the heartbeat may have been created using a fallback method or the
+	// upstream doesn't support graceful stop.
```

Comment:
Can't we get rid of the fallback announce at this point, so we'd only need to deal with upstreams not supporting in-band delete?